File: scripts/transpuller.py
#!/usr/bin/python
#https://github.com/cryptean/bitcoinlib
import sys, os, re
from bitcoinrpc.authproxy import AuthServiceProxy

import mysql.connector

RPC_ADDRESS="127.0.0.1:9443"
RPC_USER="YOUR_RPC_USER"
RPC_PASSWORD="YOUR_SUPER_SECRET_RPC_PASSWORD"

class Address:
	def __init__(self, address):
		self.address = address
		self.balance = 0
		self.sent = 0
		self.received = 0
		self.voutList = []		
		self.vinList = []
		self.inTxCount = 0
		self.outTxCount = 0
		
	def addVout(self, txid, amount):
		self.voutList.append({'amount': amount, 'txid': txid})
		self.received = self.received + amount
		self.outTxCount = self.outTxCount + 1
		self.balance = self.received - self.sent
		
	def addVin(self, txid, amount):
		self.vinList.append({'amount': amount, 'txid': txid})
		self.sent = self.sent + amount
		self.inTxCount = self.inTxCount + 1
		self.balance = self.received - self.sent
		
		if self.address == 'coinbase':
			self.balance = self.sent
		
	def getBalance(self):
		print(self.address, 'inTxCount:', self.inTxCount, 'outTxCount:', self.outTxCount, 'balance:', format(self.balance, '.8f'))
		#print('Out transactions')
		#for vou in self.voutList:
		#	print('tx', vou['txid'],'amount', vou['amount'])
			
		#print('\n')
		#print('In transactions')
		#print('\n')
		#for vin in self.vinList:
		#	print('tx', vin['txid'],'amount', vin['amount'])
			

def connect(address, user, password):
	return AuthServiceProxy("http://%s:%s@%s"%(user, password, address))
	
def extract_block_addresses(rpc, block):
	addresses = []
	misses = 0
	
	for tx in block[u'tx']:
		raw_tx = rpc.getrawtransaction(tx, True)
		
		if 'vin' in raw_tx:
			for vin in raw_tx[u'vin']:
				if 'coinbase' in vin:#coinbase transactions
					amount = 0
					for vout in raw_tx[u'vout']:
						amount = amount + vout['value']
						
					addresses.append({'address': 'coinbase', 'amount': amount, 'txid': raw_tx['txid'], 'in': 1})
				elif 'txid' in vin:#no coinbase transactions
					raw_vin_tx = rpc.getrawtransaction(vin['txid'], True)
					ref_vout = raw_vin_tx['vout'][vin['vout']]
					addresses.append({'address': ref_vout['scriptPubKey']['addresses'][0], 'amount': ref_vout['value'], 'txid': vin['txid'], 'in': 1})
		
		if 'vout' not in raw_tx:
			sys.stderr.write("Transaction %s has no 'vout': %s\n"%(tx, raw_tx))
			break
		
		for vout in raw_tx[u'vout']:
			if 'scriptPubKey' not in vout:
				sys.stderr.write("Vout %s of Transaction %s has no 'scriptPubKey'\n"%(vout, tx))
				break
			
			if vout['scriptPubKey']['type'] == 'nulldata' or vout['scriptPubKey']['type'] == 'nonstandard':
				misses = misses + 1
			elif 'addresses' in vout['scriptPubKey']:
				addresses.append({'address': vout['scriptPubKey']['addresses'][0], 'amount': vout['value'], 'txid': raw_tx['txid'], 'in': 0})
			else:
				sys.stderr.write("Can't handle %s transaction output type in transaction %s\n"%(vout["scriptPubKey"]["type"], raw_tx))
			
	
	return addresses

def main():
	print('Copyright (C) 2017 - 2019, Betchip core')
	print('\n')

	addressMap = {}
	
	if len(sys.argv) > 1:
		start_block = int(sys.argv[1])
	else:
		start_block = 1
	
	if len(sys.argv) > 2:
		end_block = int(sys.argv[2])
	else:
		end_block = 0
		
	rpc = connect(RPC_ADDRESS, RPC_USER, RPC_PASSWORD)
	if end_block == 0:
		end_block = rpc.getblockcount()
		
	firstBlock = start_block
	# Position at first block and then iterate over list using nextblockhash to save one extra call
	block_hash = rpc.getblockhash(firstBlock)
	for b in range(start_block, end_block+1):
		try:
			block = rpc.getblock(block_hash)
			
			for addr in extract_block_addresses(rpc, block):
				if addr['address'] not in addressMap:
					address = Address(addr['address'])
					
					if(addr['in'] == 1):
						address.addVin(addr['txid'], addr['amount'])
					else:
						address.addVout(addr['txid'], addr['amount'])
						
					addressMap[addr['address']] = address
				else:
					address = addressMap[addr['address']]
					if(addr['in'] == 1):
						address.addVin(addr['txid'], addr['amount'])
					else:
						address.addVout(addr['txid'], addr['amount'])
			
			if 'nextblockhash' in block:
				block_hash = block['nextblockhash']
		except:
			rpc = connect(RPC_ADDRESS, RPC_USER, RPC_PASSWORD)
			#block_hash = rpc.getblockhash(b)
			#for addr in extract_block_addresses(rpc, block_hash):
			#	print(addr)

	g = open("btpaddress.txt", "a")
	for k in addressMap:
		if addressMap[k].balance > 0:
			g.write('%s,%s\n' % (k, format(addressMap[k].balance, '.8f')))
		#print(k, 'amount:', addressMap[k]['amount'])

	#for k in addressMap:
	#	addressMap[k].getBalance()
		
if __name__ == '__main__':
	main()

File: scripts/transpuller.php
<?php
// Reference:
//	https://github.com/aceat64/EasyBitcoin-PHP/blob/master/easybitcoin.php

define("RPC_IPADDRESS", "127.0.0.1");
define("RPC_PORT", "9443");
define("RPC_USER", "YOUR_RPC_USER");
define("RPC_PASSWORD", "YOUR_SUPER_SECRET_RPC_PASSWORD");

define("DB_CONNECTION", "mysql");
define("DB_HOST", "127.0.0.1");
define("DB_PORT", "3306");
define("DB_DATABASE", "YOUR_DB");
define("DB_USERNAME", "YOUR_DB");
define("DB_PASSWORD", "YOUR_DB");

require_once('easybitcoin.php');
require_once('models.php');

function extract_block_addresses($rpc, $block){
	$addresses = [];
	
	foreach ($block['tx'] as $tx) {
		$raw_tx = $rpc->getrawtransaction($tx, true);
		
		if(isset($raw_tx['vin'])){
			foreach ($raw_tx['vin'] as $vin) {
				if(isset($vin['coinbase'])){
					$amount = 0;
					foreach ($raw_tx['vout'] as $vout){
						$amount = $amount + $vout['value'];
					}
					//echo "tx: $tx is Coinbase worth: $amount\n";
					$addresses[] = array('address'=> 'coinbase', 'amount'=> $amount, 'txid'=> $tx, 'in'=> 1, 'time'=>$raw_tx['time'], 'blockhash'=>$raw_tx['blockhash'], 'height'=>$block['height']);
				}else if(isset($vin['txid'])){
					$raw_vin_tx = $rpc->getrawtransaction($vin['txid'], true);
					$ref_vout = $raw_vin_tx['vout'][$vin['vout']];
					$addresses[] = array('address'=> $ref_vout['scriptPubKey']['addresses'][0], 'amount'=> $ref_vout['value'], 'txid'=> $raw_tx['txid'], 'in'=> 1, 'time'=>$raw_tx['time'], 'blockhash'=>$raw_tx['blockhash'], 'height'=>$block['height']);
				}
			}			
		}

		foreach ($raw_tx['vout'] as $vout) {
			if(isset($vout['scriptPubKey']) && $vout['scriptPubKey']['type'] != 'nulldata' && $vout['scriptPubKey']['type'] != 'nonstandard'){
				$addresses[] = array('address'=> $vout['scriptPubKey']['addresses'][0], 'amount'=> $vout['value'], 'txid'=> $raw_tx['txid'], 'in'=> 0, 'time'=>$raw_tx['time'], 'blockhash'=>$raw_tx['blockhash'], 'height'=>$block['height']);
			}
		}
	}
	
	return $addresses;
}

$bitcoin = new Bitcoin(RPC_USER, RPC_PASSWORD, RPC_IPADDRESS, RPC_PORT);
$settings = new Settings(DB_HOST, DB_DATABASE, DB_USERNAME, DB_PASSWORD);

$blockchaininfo = $bitcoin->getblockchaininfo();

$begin_block = $settings->getBeginBlockHeight();
$end_block = 0;

if(isset($blockchaininfo['blocks'])){
	$end_block = $blockchaininfo['blocks'];
}else{
	exit('Could not connect to RPC Node.');
}

$block_hash = $bitcoin->getblockhash($begin_block);
$block_index = $begin_block;
$addrMap = array();

while ($block_index < ($end_block+1)){
	$block = $bitcoin->getblock($block_hash);
	//echo $block_hash."\n";
	$addresses = extract_block_addresses($bitcoin, $block);
	foreach ($addresses as $addr) {
		$scriptPubKey = $addr['address'];
		if (!isset($addrMap[$scriptPubKey])){
			$addrMap[$scriptPubKey] = new Address($scriptPubKey);
		}
		
		$address = $addrMap[$scriptPubKey];
		$address->appendTransaction($addr);
	}
	
	// BEGIN
	//Let's do one block at a time to avoid big loss in case of failure
	try {
		$settings->saveAddressMap($addrMap, $block_index);
	} catch (Exception $e) {
	  error_log($e->getMessage());
	  exit('ERROR:'.$e->getMessage()); //something a user can understand
	}
	$addrMap = array();
	// END
	
	if(isset($block['nextblockhash'])){
		$block_hash = $block['nextblockhash'];
	}	
	
	$block_index++;
}